### ISSUE TITLE
Add Constraint to test whether actual item is contained in expected collection

### ISSUE THREAD

**Issue by Dreamescaper** [nunit/nunit#2704]
It would be great to have a constraint, inverse to Does.Contain(...) constraint.

Proposed syntax:
`Assert.That("1", Is.ItemFrom("1", "2", "3"))`

Why needed?
- if you use
`Assert.That(expectedValues, Does.Contain(actualValue))`
in case of failure it will log something like this:
```
Expected: does contain *actualValue*
Actual: *expectedValues*
```
Which is really confusing, since `actual` and `expected` are switched places.

- Other case - if you'd like to verify that collection consists only of items from another collection.
Currently it can be done like that:
```
var actual = new [] {1,2,3,2,3,2,2,1,1};
Assert.That(actual.Distict(), Is.SubsetOf(1,2,3))
```
which is not readable and logging would suffer.
With proposed constraint it will be:
```
var actual = new [] {1,2,3,2,3,2,2,1,1};
Assert.That(actual, Has.All.ItemFrom(1,2,3))
```
 
Initial implementation:
https://github.com/nunit/nunit/pull/2703

Suggested constraint name: 
`Is.ItemFrom(1,2,3)`, `Has.All.ItemFrom(1,2,3)`
`Is.In(1,2,3)`, `Has.All.In(1,2,3)`
`Is.AnyOf(1,2,3)`, `Has.All.AnyOf(1,2,3)`

**Comment by Dreamescaper** [comment]
@jnm2 
I like how `.AnyOf` looks better than `.In` or `.ItemFrom`, although latter ones might be easier to find using IntelliSense, if you don't know exact name (if you're just trying to find constraint, but not really sure if that exists at all).
Any of suggested names are great for me though

**Comment by jnm2** [comment]
I like this proposal because we don't currently have a concise way to do `Is.EqualTo(2).Or.EqualTo(3).Or.EqualTo(5)`.

The `params` overload suggests that you have a fixed number of items, so it's still possible that the `...Or.EqualTo(x)` method may be a good option for you. However, I agree there's really no contest versus `Is.AnyOf(1, 2, 3)` for readability.

**Comment by jnm2** [comment]
The proposed API has an issue:

```cs
AnyOf("string");

void AnyOf(params object[] x) => Console.Write("You thought you were enumerating strings...");
void AnyOf(IEnumerable x) => Console.Write("...surprise! You're enumerating chars. 👿");
```

Ideally `AnyOf("s1", "s2", "s3")`, `AnyOf("s1", "s2")`, `AnyOf("s1")` should be a consistent progression.

**Comment by Dreamescaper** [comment]
Damn, missed that :)
Even though I like `AnyOf(1,2,3)` much better than `AnyOf(new[]{ 1,2,3 })` for a case when expected options are fixed, not sure it's worth adding creepy workarounds to avoid that.

Do you know any elegant way to resolve that?

**Comment by jnm2** [comment]
@Dreamescaper Besides special-casing `string` via a dedicated overload or by type-checking at runtime, i don't know of any yet. 😢 https://gitter.im/dotnet/csharplang?at=5a7db3cd18f388e6268186ae

**Comment by jnm2** [comment]
I guess we'll have to decide which we want to live with: switching to the IEnumerable overload for types like `string` when there is only one item—maybe not the end of the world—or having to write `new[] { }` every time.
Too bad we don't ship analyzers in the main NuGet package yet. `Is.OneOf(str)` would be a perfect example of something that should be flagged as a warning with an option to replace with `EqualTo(str)`.

Since the main use case is multiple values, let's prioritize that and keep the `params` overload. There's a thought—what if we threw out the `IEnumerable` overload?

**Comment by Dreamescaper** [comment]
Well, I like having ability to provide any collection type, especially for cases when you don't know expectedItems at compile time.
I think it would be used more than .OneOf with single argument non-collection argument.

>  switching to the IEnumerable overload for types like string when there is only one item
Problem is that we need to switch _from_ IEnumerable overload.

We can do something like that
'void AnyOf(object firstExpected, params [] otherExpected)`
but not sure that's great idea :)

**Comment by jnm2** [comment]
> I think it would be used more than .OneOf with single non-collection argument.

Can't argue with that. I'm pretty close to calling that an antipattern at this point.

> `AnyOf(object firstExpected, params object[] otherExpected)`

Doesn't solve the `AnyOf("string")` problem. The problem is that anything that allows `string` to be treated as a single item by definition also allows `List<string>` to be treated as a single item.

**Comment by Dreamescaper** [comment]
Yeah.
Anyway I don't think any workarounds should be done here. 
If we strongly feel that single argument should be consistent (seems like we're not though :) ), I'd rather remove one of the overloads.

**Comment by CharliePoole** [comment]
@Dreamescaper Thanks for creating an issue... it's better to talk about __what__ we want to do on issues and only deal with implementation details on the PR.

I was initially confused on the PR but now that I understand the objective, I agree that this is needed. We should make sure the language of the user interface matches the __logical__ operation you are trying to perform rather than reflecting the implementation. In this case, the logical operation is that of equality to one of a set of things, whereas the implementation will obviously use containment.

Your alternatives `ItemFrom` and `In` speak to me about containment, not equality. Anything that's put on the `Has` class also has the same implication. The operation we actuall are trying to create is something like `EqualToAnyOf` and I can accept `AnyOf` as a reasonable short form for that. I'd go with `AnyOf``

The whole issue with params arrays is a hard one. In the past, I've resolved it by providing overloads for small numbers of entries. It seems to me, however, that over time new compilers have ended up picking different overloads, which has made maintenance tricky.

What about `AnyOf<T>( T val1, params T[] remainingVals)`? That forces all the values to be the same type and may resolve the question. If we want to disallow use of  `AnyOf`` with a single argument, we could specify the first two values instead.

**Comment by jnm2** [comment]
@CharliePoole `AnyOf<T>(T arg0, params T[] rest)` doesn't solve the problem because `.AnyOf(new List<string> { "x" })` (or more generally `.AnyOf(expectedList)`) calls that overload rather than calling `.AnyOf(IEnumerable)`.

`AnyOf<T>(T arg0, T arg1, params T[] rest)` doesn't solve the problem because `.AnyOf("x")` calls the `IEnumerable` overload. Unless that would be considered a feature. Right now that seems confusing.

Bottom line is, without special-casing, either both `string` and `List<>` will be considered collection types, or both will not. I guess people will have to learn how overload resolution works on their own since we don't ship diagnostic analyzers yet.

**Comment by Dreamescaper** [comment]
@jnm2 @CharliePoole 

Actually, simply making method generic solves the issue.
So even 
`AnyOf<T>( T val1, params T[] remainingVals)`
works just fine

**Comment by CharliePoole** [comment]
@jnm2 Why would you have an IEnumerable overload, given the intent of the feature?

**Comment by jnm2** [comment]
@CharliePoole 

Convenience in some situations? @Dreamescaper said:

> Well, I like having ability to provide any collection type, especially for cases when you don't know expectedItems at compile time.
> I think it would be used more than .OneOf with single non-collection argument.

But I guess we could just have the `params` overload and make people use `.ToArray()` for dynamic expected collections.

**Comment by CharliePoole** [comment]
A test where you don't know the expected values at compile time is suspicious. Not saying it's impossible or necessarily bad - just suspicious. After all, you are supposed to design tests for which you __know__ the intended result.

In rare cases, you can use the constraint directly, but the syntax classes are supposed to guide people to the most common, preferred use. They do that first of all via intellisense and as a fallback by compiler failures.

**Comment by jnm2** [comment]
@CharliePoole What about when expectedValues is a test method parameter?

**Comment by CharliePoole** [comment]
Yes, that's a use case I hadn't thought of. It's a problem. If we only had the `OneOf` overload I hypothesized, then I'd write the test case to take the same pattern... one object of type, followed by remaining params array of type. It's admittedly ugly though.

**Comment by jnm2** [comment]
What are the chances? I have to write: `Is.Not.EqualTo(TestMessageLevel.Error).And.Not.EqualTo(TestMessageLevel.Warning)`

and `Is.Not.AnyOf(TestMessageLevel.Error, TestMessageLevel.Warning)` looks really good right now. 😄

**Comment by CharliePoole** [comment]
My guess is that 99.9% of the usage would be just like that: a short list of scalar values.

**Comment by jnm2** [comment]
We could add the params now and do IEnumerable on demand later, maybe? The workaround would be `.AnyOf(expectedList.ToArray())`—not bad at all.

**Comment by Dreamescaper** [comment]
To sum up:
- naming: Is.AnyOf(1,2,3)
- no IEnumerable override, only params object[]

Anything else?

**Comment by CharliePoole** [comment]
Yup, We can continue with the details on the PR when you are ready.

**Comment by Dreamescaper** [comment]
Updated PR (https://github.com/nunit/nunit/pull/2703) according to discussion

**Comment by Dreamescaper** [comment]
Pr is submitted, closing the issue

**Comment by CharliePoole** [comment]
@Dreamescaper FYI we try to put "Fixes #2704" in the initial PR description so that it will autoclose.

**Comment by jnm2** [comment]
https://help.github.com/articles/closing-issues-using-keywords/
